modules/stochastic_tools/src/surrogates/BasicNNTrainer.C

```diff
@@ -0,0 +1,185 @@
+    _pcols(getParam<std::vector<unsigned int>>("predictor_cols")),
+    _response(getTrainingData<Real>(getParam<ReporterName>("response"))),
+    _no_batches(getParam<unsigned int>("no_batches")),
+    _no_epocs(getParam<unsigned int>("no_epochs")),
+    _no_hidden_layers(getParam<unsigned int>("no_hidden_layers")),
+    _no_neurons_per_layer(getParam<std::vector<unsigned int>>("no_neurons_per_layer")),
+    _learning_rate(getParam<Real>("learning_rate"))
+{
+  const auto & pnames = getParam<std::vector<ReporterName>>("predictors");
+  for (unsigned int i = 0; i < pnames.size(); ++i)
+    _pvals[i] = &getTrainingData<Real>(pnames[i]);
+
+  // If predictors and predictor_cols are empty, use all sampler columns
+  if (_pvals.empty() && _pcols.empty())
+  {
+    _pcols.resize(_sampler.getNumberOfCols());
+    std::iota(_pcols.begin(), _pcols.end(), 0);
+  }
+
+  // Resize sample points to number of predictors
+  _sample_points.resize(_pvals.size() + _pcols.size() + 1);
+
+  if (_no_hidden_layers != _no_neurons_per_layer.size())
+    mooseError("The number of layers are not the same!");
+}
+
+void
+BasicNNTrainer::preTrain()
+{
+  // Resize to number of sample points
+  for (auto & it : _sample_points)
+    it.resize(_sampler.getNumberOfLocalRows());
+}
+
+void
+BasicNNTrainer::train()
+{
+  unsigned int d = 0;
+  // Get predictors from reporter values
+  for (const auto & val : _pvals)
+    _sample_points[d++][_local_row] = *val;
+  // Get predictors from sampler
+  for (const auto & col : _pcols)
+    _sample_points[d++][_local_row] = _sampler_row[col];
+
+  _sample_points.back()[_local_row] = _response;
+}
+
+void
+BasicNNTrainer::postTrain()
+{
+  for (auto & it : _sample_points)
+    _communicator.allgather(it);
+
+#ifdef TORCH_ENABLED
+
+  // Fixing the RNG seed to make sure every experiment is the same.
+  // Otherwise sampling / stochastic gradient descent would be different.
+  torch::manual_seed(11);
+
+  // Now we convert the internal data-structure to torch::Tensors
+  std::vector<Real> flattened_data;
+  std::vector<Real> flattened_response;
+
+  // First, we flatten (serialize) out containers
+  for (const auto & r : _sample_points)
+  {
+    if (&r == &_sample_points.back())
+      for (const auto & c : r)
+        flattened_response.push_back(c);
+    else
+      for (const auto & c : r)
+        flattened_data.push_back(c);
+  }
+
+  // Then, we create and load our Tensors
+  int n_rows = _sample_points.size() - 1;
+  int n_cols = _sample_points[0].size();
+
+  // The default data type in pytorch is float, while we use double in MOOSE.
+  // Therefore, in some cases we have to convert Tensors to double.
+  auto options = torch::TensorOptions().dtype(at::kDouble);
+  torch::Tensor data_tensor =
+      torch::from_blob(flattened_data.data(), {n_rows, n_cols}, options).to(at::kDouble);
+  torch::Tensor response_tensor =
+      torch::from_blob(flattened_response.data(), {n_cols}, options).to(at::kDouble);
```

Comment:
In fact, is there a way to fill in these tensors in `train`? Not sure how the mpi would work though.

Reply:
We are filling these in train() in the latest version. Thanks for the great suggestion!